libnetwork/ipam/parallel_test.go

```diff
@@ -223,7 +223,7 @@ func allocate(t *testing.T, tctx *testContext, parallel int64) {
 			continue
 		}
 		if there, ok := tctx.ipMap[ip.String()]; ok && there {
-			t.Fatalf("Got duplicate IP %s", ip.String())
+			t.Errorf("Got duplicate IP %s", ip.String())
 			break
```

Comment:
Thanks for contributing (sorry for the delay, I saw your PR, but wanted to have a slightly closer look).

So, there's indeed unreachable code in this function (the `break` would never be reached), but I'm not sure if changing `t.Fatal` to `t.Error` is the correct fix here. i'm not _very_ familiar with this test, but looking at the code;

- `allocate()` is used by various tests
- all of them look to be using it to run these steps _in parallel_

Changing to `t.Error()` effectively would mean;

- All parallel runs continue running
- This `for` loop also continues running
- (probably) the code below the loop would still `Fatal` the test


So the "old" code (before this PR) would fail early; based on all of the above, I _think_ that's intentional, because if one of the (parallel) runs fails, the test should fail, so there's not really a reason to continue running the other (parallel) tests as they're not really testing anything else.

I would suggest to instead keep the `t.Error()`, and remove the `break`;

```go
if there, ok := tctx.ipMap[ip.String()]; ok && there {
	t.Fatalf("Got duplicate IP %s", ip.String())
}
```

While updating, perhaps also change the code outside of the loop, because the `assert.Check` is redundant, and;

```go
assert.Check(t, is.Len(tctx.ipList, tctx.maxIP))
if len(tctx.ipList) != tctx.maxIP {
	t.Fatal("mismatch number allocation")
}
```


Can be changed to an `assert.Assert()` which would check the length and call `t.FailNow()` if it fails;

```go
assert.Assert(t, is.Len(tctx.ipList, tctx.maxIP))
```

Reply:
Thank you very much for your patient review and suggestions, the code has been updated.